src/osd/OSD.cc

```diff
@@ -9567,112 +9224,324 @@ int OSD::init_op_flags(OpRequestRef& op)
+	p = pg_slots.erase(p);
+	continue;
       }
-      p->second.to_process.clear();
-      p->second.waiting_for_pg = false;
-      ++p->second.requeue_seq;
-      queued = true;
     }
-  }
-  if (pushes_to_free > 0) {
-    osd->service.release_reserved_pushes(pushes_to_free);
+    ++p;
   }
   if (queued) {
-    sdata->sdata_lock.Lock();
-    sdata->sdata_cond.SignalOne();
-    sdata->sdata_lock.Unlock();
+    sdata_lock.Lock();
+    sdata_cond.SignalOne();
+    sdata_lock.Unlock();
+  }
+}
+
+void OSDShard::_wake_pg_slot(
+  spg_t pgid,
+  OSDShardPGSlot *slot)
+{
+  dout(20) << __func__ << " " << pgid
+	   << " to_process " << slot->to_process
+	   << " waiting " << slot->waiting
+	   << " waiting_peering " << slot->waiting_peering << dendl;
+  for (auto i = slot->to_process.rbegin();
+       i != slot->to_process.rend();
+       ++i) {
+    _enqueue_front(std::move(*i), osd->op_prio_cutoff);
+  }
+  slot->to_process.clear();
+  for (auto i = slot->waiting.rbegin();
+       i != slot->waiting.rend();
+       ++i) {
+    _enqueue_front(std::move(*i), osd->op_prio_cutoff);
+  }
+  slot->waiting.clear();
+  for (auto i = slot->waiting_peering.rbegin();
+       i != slot->waiting_peering.rend();
+       ++i) {
+    // this is overkill; we requeue everything, even if some of these items are
+    // waiting for maps we don't have yet.  FIXME.
+    for (auto j = i->second.rbegin(); j != i->second.rend(); ++j) {
+      _enqueue_front(std::move(*j), osd->op_prio_cutoff);
+    }
   }
+  slot->waiting_peering.clear();
+  slot->waiting_for_split = false;
+  ++slot->requeue_seq;
 }
 
-void OSD::ShardedOpWQ::prune_pg_waiters(OSDMapRef osdmap, int whoami)
+void OSDShard::identify_splits(OSDMapRef as_of_osdmap, set<spg_t> *pgids)
 {
-  unsigned pushes_to_free = 0;
-  for (auto sdata : shard_list) {
-    Mutex::Locker l(sdata->sdata_op_ordering_lock);
-    sdata->waiting_for_pg_osdmap = osdmap;
-    auto p = sdata->pg_slots.begin();
-    while (p != sdata->pg_slots.end()) {
-      ShardData::pg_slot& slot = p->second;
-      if (!slot.to_process.empty() && slot.num_running == 0) {
-	if (osdmap->is_up_acting_osd_shard(p->first, whoami)) {
-	  dout(20) << __func__ << "  " << p->first << " maps to us, keeping"
-		   << dendl;
-	  ++p;
-	  continue;
-	}
-	while (!slot.to_process.empty() &&
-	       slot.to_process.front().get_map_epoch() <= osdmap->get_epoch()) {
-	  auto& qi = slot.to_process.front();
-	  dout(20) << __func__ << "  " << p->first
-		   << " item " << qi
-		   << " epoch " << qi.get_map_epoch()
-		   << " <= " << osdmap->get_epoch()
-		   << ", stale, dropping" << dendl;
-	  pushes_to_free += qi.get_reserved_pushes();
-	  slot.to_process.pop_front();
-	}
-      }
-      if (slot.to_process.empty() &&
-	  slot.num_running == 0 &&
-	  !slot.pg) {
-	dout(20) << __func__ << "  " << p->first << " empty, pruning" << dendl;
-	p = sdata->pg_slots.erase(p);
+  Mutex::Locker l(sdata_op_ordering_lock);
+  if (osdmap) {
+    for (auto& i : pg_slots) {
```

Comment:
This ```identify_splits()``` implementation *only* exists to close a very narrow race in "pg query", right? I'm not sure that's worth the tradeoff of extra iterations on every OSDMap update, although I'm not opposed to it either.

Reply:
It's the general pg slot priming that makes any event wait for the split to be processed and child pg to be instantiated.  I suspect it will affect lots more than just query... I didn't spend a lot of time analyzing what a sloppy approach to the slot tracking would affect, though!